src/Microsoft.AspNet.Mvc.Core/Infrastructure/ObjectResultExecutor.cs

```diff
@@ -256,19 +256,11 @@ public virtual Task ExecuteAsync(ActionContext context, ObjectResult result)
             HttpRequest request)
         {
             var result = new List<MediaTypeSegmentWithQuality>();
-            var parsedHeaders = request.GetTypedHeaders().Accept;
-            for (var i = 0; i < parsedHeaders?.Count; i++)
+            AcceptHeaderParser.ParseAcceptHeader(request.Headers[HeaderNames.Accept], result);
+            for (int i = 0; i < result.Count; i++)
             {
-                result.Add(new MediaTypeSegmentWithQuality(
-                    new StringSegment(parsedHeaders[i].ToString()),
-                    parsedHeaders[i].Quality ?? 1.0));
-            }
-
-            for (var i = 0; i < result.Count; i++)
-            {
-                if (!RespectBrowserAcceptHeader &&
-                    MediaTypeComparisons.MatchesAllTypes(result[i].MediaType) &&
-                    MediaTypeComparisons.MatchesAllSubtypes(result[i].MediaType))
+                var mediaType = new MediaType(result[i].MediaType);
```

Comment:
MediaType is a heavy struct, if we have a list of MediaType it will show up in the profile in a negative way, so I think it's best to have a separate type to represent just the minimum amount of information that we care about at this stage and use MediaType locally only for doing comparisons (like is subset of, matches all types, etc)